### PR TITLE
[SPARK-40737][CONNECT] Add basic support for DataFrameWriter

### PR DESCRIPTION
**Pull request by grundprinzip**
### What changes were proposed in this pull request?

This change adds basic support for writes through the Spark Connect API. In todays implementation of the write API from a DataFrame perspective the interface of the DataFrameWriter is as declarative as possible today. 

The write support is implemented as a `Command` and does not return anything.

### Why are the changes needed?

Write support through Spark Connect.

### Does this PR introduce _any_ user-facing change?
Experimental API

### How was this patch tested?

Added new unit tests for the behavior.

### CODE REVIEW

File: connector/connect/src/main/protobuf/spark/connect/commands.proto [hunk 1]
```diff
@@ -62,3 +65,29 @@ message CreateScalarFunction {
     FUNCTION_LANGUAGE_SCALA = 3;
   }
 }
+
+// As writes are not directly handled during analysis and planning, they are modeled as commands.
```
Comment (hvanhovell) [inline connector/connect/src/main/protobuf/spark/connect/commands.proto:69]: We want this to be eagerly executed, so - IMO - making this a command is your only option.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala [hunk 2]
```diff
@@ -74,4 +77,49 @@ class SparkConnectCommandPlanner(session: SparkSession, command: proto.Command)
...
+  /**
+   * Transforms the write operation and executes it.
+   *
+   * The input write operation contains a reference to the input plan and transforms it to
+   * the corresponding logical plan. Afterwards, creates the DataFrameWriter and translates
+   * the parameters of the WriteOperation into the corresponding methods calls.
+   *
+   * @param writeOperation
+   */
+  def handleWriteOperation(writeOperation: WriteOperation): Unit = {
```
Comment (hvanhovell) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:89]: It is a bit weird to have this in the SparkPlanner node, but I guess this is the consequence of the builder() API we have in the DataFrameWriter.

@cloud-fan AFAIK you have been working on making writes more declarative (i.e. planned writes). Do you see a way to improve this?

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connect.planner
+
+import org.apache.spark.SparkFunSuite
+
+class SparkConnectCommandPlannerSuite extends SparkFunSuite with SparkConnectPlanTest {
+
+  test("Write Relation") {
```
Comment (hvanhovell) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala:24]: We probably need a little bit more coverage :)

Comment (grundprinzip) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala]: I added a bunch of more tests for the command translation.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 4]
```diff
@@ -23,7 +23,13 @@ import org.apache.spark.annotation.{Since, Unstable}
 import org.apache.spark.connect.proto
 import org.apache.spark.sql.SparkSession
 import org.apache.spark.sql.catalyst.{expressions, plans}
-import org.apache.spark.sql.catalyst.analysis.{UnresolvedAlias, UnresolvedAttribute, UnresolvedFunction, UnresolvedRelation, UnresolvedStar}
+import org.apache.spark.sql.catalyst.analysis.{
```
Comment (grundprinzip) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:26]: fix

**Comment by grundprinzip** [comment]
@cloud-fan @amaliujia @hvanhovell please take a look!

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala [hunk 5]
```diff
@@ -24,10 +24,16 @@ import com.google.common.collect.{Lists, Maps}
...
-import org.apache.spark.sql.SparkSession
+import org.apache.spark.connect.proto.WriteOperation
+import org.apache.spark.sql.{Dataset, SparkSession}
+import org.apache.spark.sql.connect.planner.SparkConnectPlanner
 import org.apache.spark.sql.execution.python.UserDefinedPythonFunction
 import org.apache.spark.sql.types.StringType
 
+final case class InvalidCommandInput(
+    private val message: String = "",
+    private val cause: Throwable = None.orNull)
```
Comment (hvanhovell) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:35]: use `null`?

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala [hunk 6]
```diff
@@ -24,10 +24,16 @@ import com.google.common.collect.{Lists, Maps}
...
 import org.apache.spark.api.python.{PythonEvalType, SimplePythonFunction}
 import org.apache.spark.connect.proto
-import org.apache.spark.sql.SparkSession
+import org.apache.spark.connect.proto.WriteOperation
+import org.apache.spark.sql.{Dataset, SparkSession}
+import org.apache.spark.sql.connect.planner.SparkConnectPlanner
 import org.apache.spark.sql.execution.python.UserDefinedPythonFunction
 import org.apache.spark.sql.types.StringType
 
+final case class InvalidCommandInput(
```
Comment (hvanhovell) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:33]: Should we use `IllegalArgumentException` here? Or do you feel this needs its own specific exception?

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala [hunk 7]
```diff
@@ -74,4 +82,60 @@ class SparkConnectCommandPlanner(session: SparkSession, command: proto.Command)
...
+      }
+      w.bucketBy(op.getBucketCount, cols.head, cols.tail.toSeq: _*)
+    }
+
+    if (writeOperation.getPartitionByColumnsCount > 0) {
+      val names = writeOperation.getPartitionByColumnsList.asScala
+      w.partitionBy(names.toSeq: _*)
+    }
+
+    if (writeOperation.getFormat != null) {
```
Comment (amaliujia) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:126]: yeah I found for `string` there was no `hasXXX` call thus we probably need this.

File: connector/connect/src/main/protobuf/spark/connect/commands.proto [hunk 8]
```diff
@@ -62,3 +65,29 @@ message CreateScalarFunction {
...
+
+  oneof save_type {
+    string path = 3;
+    string table_name = 4;
+  }
+  string mode = 5;
+  repeated string sortColumnNames = 6;
+  repeated string partitionByColumns = 7;
+  BucketBy bucketBy = 8;
+  repeated WriteOptions options = 9;
```
Comment (amaliujia) [inline connector/connect/src/main/protobuf/spark/connect/commands.proto:82]: Why just not a map<string, string>? Does proto map have a iterator (if so in later use it is the same as `repeated`)

Comment (grundprinzip) [reply connector/connect/src/main/protobuf/spark/connect/commands.proto:82]: So the interesting question is, options at least in the Scala source supports different types. It should probably be a `map<string, expressions.Literal>` but I was too lazy. I can change it.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala (hunk 7 above)
Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:126]: yes, sorry, `hasXXX` exists only for composite objects like messages but not for scalar values.

File: connector/connect/src/main/protobuf/spark/connect/commands.proto (hunk 8 above)
Comment (grundprinzip) [reply connector/connect/src/main/protobuf/spark/connect/commands.proto]: I mapped this to `map<string, string>` because behind this scenes it's converted to a string anyways.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala [hunk 9]
```diff
@@ -74,4 +82,64 @@ class SparkConnectCommandPlanner(session: SparkSession, command: proto.Command)
...
+    }
+
+    writeOperation.getSaveTypeCase match {
+      case proto.WriteOperation.SaveTypeCase.PATH => w.save(writeOperation.getPath)
+      case proto.WriteOperation.SaveTypeCase.TABLE_NAME =>
+        w.saveAsTable(writeOperation.getTableName)
+      case _ =>
+        throw new UnsupportedOperationException(
+          s"WriteOperation:SaveTypeCase not supported "
+            + "${writeOperation.getSaveTypeCase.getNumber}")
```
Comment (hvanhovell) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:141]: Number? Can we try to show the name?

Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:141]: In case the proto does not have the enum value there is no name.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala (hunk 6 above)
Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:33]: I wanted to have a custom exception for when we rethrow.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala (hunk 5 above)
Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:35]: done.

**Comment by grundprinzip** [comment]
@hvanhovell @cloud-fan @HyukjinKwon can you please have a look?

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala (hunk 6 above)
Comment (HyukjinKwon) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:33]: If this is a user-facing error, we should actually leverage errorframe work we have .. cc @gengliangwang @MaxGekk @itholic

Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:33]: I'm happy to fix this as a follow up, does it make sense?

The errors are reported back through grpc. If you point me to the right base class I can fix it then.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala (hunk 9 above)
Comment (HyukjinKwon) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:141]:
```suggestion
          "WriteOperation:SaveTypeCase not supported "
            + s"${writeOperation.getSaveTypeCase.getNumber}")
```

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala [hunk 10]
```diff
@@ -34,59 +36,109 @@ package object dsl {
...
-        proto.Expression.Alias.newBuilder().setName(alias).setExpr(expr)).build()
-
-      def < (other: proto.Expression): proto.Expression =
-        proto.Expression.newBuilder().setUnresolvedFunction(
-          proto.Expression.UnresolvedFunction.newBuilder()
-            .addParts("<")
-            .addArguments(expr)
-            .addArguments(other)
-        ).build()
+      def as(alias: String): proto.Expression = proto.Expression
```
Comment (HyukjinKwon) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:50]: qq is this just a style change? Maybe it's better to exclude these unrelated changes.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala [hunk 11]
```diff
@@ -40,17 +46,19 @@ class SparkConnectCommandPlanner(session: SparkSession, command: proto.Command)
...
    * This is a helper function that registers a new Python function in the SparkSession.
    *
-   * Right now this function is very rudimentary and bare-bones just to showcase how it
-   * is possible to remotely serialize a Python function and execute it on the Spark cluster.
-   * If the Python version on the client and server diverge, the execution of the function that
-   * is serialized will most likely fail.
+   * Right now this function is very rudimentary and bare-bones just to showcase how it is
+   * possible to remotely serialize a Python function and execute it on the Spark cluster. If the
+   * Python version on the client and server diverge, the execution of the function that is
+   * serialized will most likely fail.
```
Comment (HyukjinKwon) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:61]: I think we should better exclude these changes - I suspect they are from scalafmt (?).

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala (hunk 2 above)
Comment (HyukjinKwon) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:94]: cc @allisonwang-db FYI

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+   * Returns a unique valid table name indentifier on each invocation.
+   * @return
+   */
+  private def table(): String = s"table${UUID.randomUUID().toString.replace("-", "")}"
+
+  /**
+   * Helper method that takes a closure as an argument to handle cleanup of the resource created.
+   * @param thunk
+   */
+  def withTable(thunk: String => Any): Unit = {
```
Comment (HyukjinKwon) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala:57]: I think we can leverage `withTempPath` at `SQLHelper`

Comment (HyukjinKwon) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala:57]: and `withTable` at `SQLTestUtils`

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+            bucketByCols = Seq("id"),
+            numBuckets = Some(10))))
+  }
+
+  test("Write to Table") {
+    withTable { name =>
+      val cmd = localRelation.write(format = Some("parquet"), tableName = Some(name))
+      transform(cmd)
+      // Check that we can find and drop the table.
+      spark.sql(s"select count(*) from ${name}").collect()
```
Comment (HyukjinKwon) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala:153]: Should we compare the results? Could leverage `checkAnswer` at `QueryTest`

File: connector/connect/src/main/protobuf/spark/connect/commands.proto [hunk 14]
```diff
@@ -62,3 +65,39 @@ message CreateScalarFunction {
...
+  // The destination of the write operation must be either a path or a table.
+  oneof save_type {
+    string path = 3;
+    string table_name = 4;
+  }
+  SaveMode mode = 5;
+  // List of columns to sort the output by.
+  repeated string sortColumnNames = 6;
+  // List of columns for partitioning.
+  repeated string partitionByColumns = 7;
```
Comment (HyukjinKwon) [inline connector/connect/src/main/protobuf/spark/connect/commands.proto:84]: no biggie but I would match the name as what DataFrameWriter has. e.g.) `partitioningColumns`, `bucketColumnNames`, `sortColumnNames` and `extraOptions`.

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala (hunk 13 above)
Comment (grundprinzip) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala:153]: The table is empty and the assumption is that the write operation was successful if the table can be found, because the write is already implemented and tested.

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala (hunk 12 above)
Comment (grundprinzip) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala:57]: Fixing that.

File: connector/connect/src/main/protobuf/spark/connect/commands.proto (hunk 14 above)
Comment (grundprinzip) [reply connector/connect/src/main/protobuf/spark/connect/commands.proto:84]: will do

Comment (grundprinzip) [reply connector/connect/src/main/protobuf/spark/connect/commands.proto:84]: I changed this accordingly with a minor exception on the bucketing where I wrap this into the `BucketBy` message to have a better ux for setting the two values together.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala (hunk 10 above)
Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:50]: It's an annoying mix. The conflict resolution on the merge conflict was ugly and then I had to clean it up. Doing this manually sucks.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala (hunk 11 above)
Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala]: done

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectPlannerSuite.scala [hunk 15]
```diff
@@ -88,16 +108,20 @@ class SparkConnectPlannerSuite extends SparkFunSuite with SparkConnectPlanTest {
...
         .setInput(proto.Relation.newBuilder().setRead(readWithTable).build())
         .addExpressions(
-          proto.Expression.newBuilder()
-            .setUnresolvedStar(UnresolvedStar.newBuilder().build()).build()
-        ).build()
+          proto.Expression
+            .newBuilder()
+            .setUnresolvedStar(UnresolvedStar.newBuilder().build())
+            .build())
+        .build()
```
Comment (HyukjinKwon) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectPlannerSuite.scala:124]: Maybe we should get rid of these changes too?

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+
+  test("Write fails with unknown table - AnalysisException") {
+    val cmd = readRel.write(tableName = Some("dest"))
+    assertThrows[AnalysisException] {
+      transform(cmd)
+    }
+  }
+
+  test("Write with partitions") {
+    val name = table()
```
Comment (HyukjinKwon) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala:68]: maybe `withTable`

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+   * Returns a unique path name on every invocation.
+   * @return
+   */
+  private def path(): String = s"/tmp/${UUID.randomUUID()}"
+
+  /**
+   * Returns a unique valid table name indentifier on each invocation.
+   * @return
+   */
+  private def table(): String = s"table${UUID.randomUUID().toString.replace("-", "")}"
```
Comment (HyukjinKwon) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala:48]: I believe we don't need this because the tests won't run in parallel, and `withTable` will drop the table.

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+    with SparkConnectPlanTest
+    with SharedSparkSession {
+
+  lazy val localRelation = createLocalRelationProto(Seq($"id".int))
+
+  /**
+   * Returns a unique path name on every invocation.
+   * @return
+   */
+  private def path(): String = s"/tmp/${UUID.randomUUID()}"
```
Comment (HyukjinKwon) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala:42]: This one too. I believe we don't need a random path now. We could just inline any path into the test

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala [hunk 19]
```diff
@@ -34,59 +36,109 @@ package object dsl {
...
+        writeOp.setInput(logicalPlan)
+        proto.Command.newBuilder().setWriteOperation(writeOp.build()).build()
+      }
+    }
   }
 
   object plans { // scalastyle:ignore
     implicit class DslLogicalPlan(val logicalPlan: proto.Relation) {
       def select(exprs: proto.Expression*): proto.Relation = {
-        proto.Relation.newBuilder().setProject(
```
Comment (HyukjinKwon) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:67]: I would remove these style changes since the previous codes already follow the style guides we have.

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise from my end.

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala (hunk 16 above)
Comment (grundprinzip) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala:68]: I'll refactor the cases so that when we use names and errors it doesnt matter.

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectPlannerSuite.scala (hunk 15 above)
Comment (grundprinzip) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectPlannerSuite.scala:124]: Ok

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala (hunk 17 above)
Comment (grundprinzip) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala:48]: Do we not run any test in parallel? Is this guaranteed?

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala (hunk 18 above)
Comment (grundprinzip) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala:42]: ok

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala (hunk 17 above)
Comment (grundprinzip) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectCommandPlannerSuite.scala:48]: removing it

**Comment by hvanhovell** [comment]
Merging this one.

File: connector/connect/src/main/protobuf/spark/connect/commands.proto [hunk 20]
```diff
@@ -62,3 +65,39 @@ message CreateScalarFunction {
...
   }
 }
+
+// As writes are not directly handled during analysis and planning, they are modeled as commands.
+message WriteOperation {
+  // The output of the `input` relation will be persisted according to the options.
+  Relation input = 1;
+  // Format value according to the Spark documentation. Examples are: text, parquet, delta.
+  string source = 2;
+  // The destination of the write operation must be either a path or a table.
```
Comment (cloud-fan) [inline connector/connect/src/main/protobuf/spark/connect/commands.proto:75]: in DF API, people can do `df.write.format("jdbc").option("table", ...).save()` , so the destination is neither path nor table. I think an optional table name is sufficient. If table name is not given, the destination will be figured out from write options (path is just one write option).

File: connector/connect/src/main/protobuf/spark/connect/commands.proto [hunk 21]
```diff
@@ -62,3 +65,39 @@ message CreateScalarFunction {
...
+  oneof save_type {
+    string path = 3;
+    string table_name = 4;
+  }
+  SaveMode mode = 5;
```
Comment (cloud-fan) [inline connector/connect/src/main/protobuf/spark/connect/commands.proto:80]: We added `DataFrameWriterV2` because we believe `SaveMode` is a bad design. It's confusing if we write to a table, and the table location is not empty. It's unclear if the save mode should apply to table existence or location existence.

Anyway, we need to support save mode in the proto definition to support the existing DF API. If we want to support `DataFrameWriterV2` in Spark connect client, we should probably have a new proto definition without save mode.

File: connector/connect/src/main/protobuf/spark/connect/commands.proto [hunk 22]
```diff
@@ -62,3 +65,39 @@ message CreateScalarFunction {
...
+  // List of columns to sort the output by.
+  repeated string sort_column_names = 6;
```
Comment (cloud-fan) [inline connector/connect/src/main/protobuf/spark/connect/commands.proto:82]: This should be part of the BucketBy

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala (hunk 2 above)
Comment (cloud-fan) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/command/SparkConnectCommandPlanner.scala:94]: This is more than planned write. We need to create a logical plan for DF write, instead of putting implementation code in DF write APIs.
